### PR TITLE
Migrated from Newtonsoft.JSON to System.Text.Json

### LINKED ISSUE #21: Replace Newtonsoft.Json with System.Text.Json
As new projects use System.Text.Json by default, this should be our default to prevent adding extra dependencies for our users.

### PR DESCRIPTION
**Pull request by SuNNjek**
Fixes #21

As pointed out in the issue, System.Text.Json doesn't have an equivalent to the JsonObjectAttribute, so I removed that bit in the code and the corresponding tests.

### CODE REVIEW

File: src/Saunter/AsyncApiSchema/v2/Identifier.cs [hunk 1]
```diff
@@ -19,25 +19,24 @@ public override string ToString()
             return value.ToString();
         }
 
-        public class JsonConverter : Newtonsoft.Json.JsonConverter
+        public class JsonConverter : System.Text.Json.Serialization.JsonConverter<Identifier>
```
Comment (tehmantra) [inline src/Saunter/AsyncApiSchema/v2/Identifier.cs:22]: We need to register this converter by adding the attribute to the Identifier class:

```csharp
[System.Text.Json.Serialization.JsonConverter(typeof(Identifier.JsonConverter))]
public class Identifier
{
    ...
```

**Review by tehmantra — CHANGES_REQUESTED** [review]
Looks like we need to do something to handle Dictionary with a non-string key too...
https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-converters-how-to#support-dictionary-with-non-string-key


```
System.NotSupportedException: The collection type 'System.Collections.Generic.Dictionary`2[Saunter.AsyncApiSchema.v2.ServersFieldName,Saunter.AsyncApiSchema.v2.Server]' is not supported.
   at System.Text.Json.JsonClassInfo.GetElementType(Type propertyType, Type parentType, MemberInfo memberInfo, JsonSerializerOptions options)
   at System.Text.Json.JsonClassInfo.CreateProperty(Type declaredPropertyType, Type runtimePropertyType, Type implementedPropertyType, PropertyInfo propertyInfo, Type parentClassType, JsonConverter converter, JsonSerializerOptions options)
   at System.Text.Json.JsonClassInfo.AddProperty(Type propertyType, PropertyInfo propertyInfo, Type classType, JsonSerializerOptions options)
   at System.Text.Json.JsonClassInfo..ctor(Type type, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializerOptions.GetOrAddClass(Type classType)
   at System.Text.Json.WriteStackFrame.Initialize(Type type, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializer.WriteCore(Utf8JsonWriter writer, Object value, Type type, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializer.WriteCore(PooledByteBufferWriter output, Object value, Type type, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializer.WriteCoreString(Object value, Type type, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializer.Serialize[TValue](TValue value, JsonSerializerOptions options)
   at Saunter.AsyncApiMiddleware.RespondWithAsyncApiSchemaJson(HttpResponse response, AsyncApiDocument asyncApiSchema) in C:\Users\wilds\src\github.com\SuNNjek\saunter\src\Saunter\AsyncApiMiddleware.cs:line 36
   at Saunter.AsyncApiMiddleware.Invoke(HttpContext context, IAsyncApiDocumentProvider asyncApiDocumentProvider) in C:\Users\wilds\src\github.com\SuNNjek\saunter\src\Saunter\AsyncApiMiddleware.cs:line 31
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

**Comment by SuNNjek** [comment]
Hey, so I've been struggling for a bit with the Dictionary converter tbh. Serialization works fine but not deserialization (because I would need to convert the string back to a class instance somehow). I could theoretically call the constructor with the value returned by ToString() but I don't know if this would be ok. I don't know if deserialization is needed though, if not I could just throw an exception here.

It seems Microsoft fixed this in version 5 (https://github.com/dotnet/runtime/issues/30524) but this is currently only available as a prerelease.

**Comment by tehmantra** [comment]
We only need serialization for now. Fine to throw `NotSupportedException` on Deserialize

File: src/Saunter/Utils/DictionaryKeyToStringConverter.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                culture: null);
+
+            return converter;
+        }
+
+        private class DictionaryKeyToStringConverterInner<TKey, TValue> :
+            JsonConverter<IDictionary<TKey, TValue>>
+        {
+            private readonly JsonConverter<TValue> _valueConverter;
+            private readonly Type _valueType;
```
Comment (tehmantra) [inline src/Saunter/Utils/DictionaryKeyToStringConverter.cs:43]: This `_valueType` field is unused.

**Comment by tehmantra** [comment]
It seems that System.Text.Json takes a very different approach to Polymorphic serialization than Newtonsoft.Json.
https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-how-to#serialize-properties-of-derived-classes

> Serialization of a polymorphic type hierarchy is not supported. For example, if a property is defined as an interface or an abstract class, only the properties defined on the interface or abstract class are serialized, even if the runtime type has additional properties.

We have quite a few empty interfaces to essentially allow "union types" e.g. 

```csharp
interface ISchema {}

class Schema : ISchema {
    // properties ...
}

class Reference : ISchema {
    // properties ...
}
```

We will need to add custom converters for those too
```csharp
public interface ISchema {

	public class JsonConverter : JsonConverter<ISchema>
	{
		public override ISchema Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) => throw new NotSupportedException();

		public override void Write(Utf8JsonWriter writer, ISchema value, JsonSerializerOptions options)
		{
			switch (value) 
			{
				case Schema s: JsonSerializer.Serialize(writer, s, options); break;
				case Reference r: JsonSerializer.Serialize(writer, r, options); break;
				default: writer.WriteNullValue(); break;
			};
		}
	}
}
```

* `ISchema`
* `IChannelBinding`
* `IMessageBinding`
* `IOperationBinding`
* `IServerBinding`
* `IMessage`
* `IMessageTrait`
* `IOperationTrait`

**Comment by tehmantra** [comment]
Some of the enums are still annotated with `[EnumMember]`. This is no longer being picked up with System.Text.Json.
This is not a problem for most enums as they just use the camelCase string of the enum value. However there are 2 exceptions to the rule:
```csharp
public enum AsyncApiVersionString
{
    [EnumMember(Value = "2.0.0")]
    v2,
}

public enum SecuritySchemeType
{
    [EnumMember(Value = "X509")] // for some reason this uses a capital "X"
    X509,

    // other values are fine
```
https://www.asyncapi.com/docs/specifications/2.0.0#securitySchemeObject

We may need custom converters for these two

**Comment by SuNNjek** [comment]
Unfortunately you cannot apply the JsonConverterAttribute to interfaces, so I now created a converter which converts interfaces using their runtime types and added them to the serialization settings in the middleware. This makes it apply to all interfaces however (don't know if that is a problem or not).

**Review by tehmantra — APPROVED** [review]

**Comment by tehmantra** [comment]
Thanks for your perseverance on this one! It's been a larger change than expected.

**Pull request merged**